### ISSUE TITLE
Support labels in template string

### ISSUE THREAD

**Issue by JesusTheHun** [microsoft/TypeScript#57385]
### 🔍 Search Terms

template string label
template string name
interpolation label

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

The idea is to allow the addition of labels into a template literal.
This has no effect on the type itself in any way, the sole purpose is to give a better DX.

These labels would be shown by IntelliSense in order to provide the same experience as labels for the tuple labels.

### 📃 Motivating Example

```ts
// Template literal today
type Options = {
   name: `${string}/${string}`;
}

// With labels
type Options = {
   name: `${scope: string}/${baseName: string}`;
}
```

### 💻 Use Cases

This would offer a slightly better DX. 
Right now you can document the parts of the template in a JSDoc comment, but the JSDoc is not preserve by IntelliSense in some cases.

**Comment by fatcerberus** [comment]
How would this work with literal placeholders?  For example `${"a" | "b"}${"a" | "b"}` expands to `"aa" | "ab" | "ba" | "bb"`

**Comment by JesusTheHun** [comment]
> How would this work with literal placeholders? For example `${"a" | "b"}${"a" | "b"}` expands to `"aa" | "ab" | "ba" | "bb"`

The label is always placed first within the brackets and outside of any quote of any kind  :
```ts
type LabelledLiteralPlaceholder = `${firstLetter: "a" | "b"}${lastLetter: "a" | "b"}`;
```

**Comment by fatcerberus** [comment]
No, I got that; I just meant the label couldn't be preserved with the type that case because the template would just eagerly expand to a union of string literals.  So half the time you'd have no way of observing the labels like you can with tuple types.

**Comment by JesusTheHun** [comment]
I guess it depends on your case. 

In my experience it never occurred that I wanted a label for static part.
